Activate roles by affinity, then by the least number of active roles

jrafanie opened the pull request:
Part of the fix for #11329

Some roles are only activated on one appliance in the region or zone.  When
multiple servers has such a role enabled (assigned), a decision has to be made
as to which server will activate the role.  Previously, this decision only
checked server affinity(priority), which allows some roles to be 'pinned' to a
specific server.

We need to honor server affinity (priority) first, but next, we can choose to
activate roles on the server with the least active roles.

Note, if a new server becomes available with the same role affinity (priority)
after we activate roles, we don't shift roles to this new server to evenly
distribute the work.

Before this fix, the attached tests fail, demonstrating that role activation
would not evenly distribute roles:

```
  1) Server Monitor After Setup, with 2 Servers in 2 Zones where I am the Master should evenly distribute singular roles on multiple servers when migrating roles from low priority server
     Failure/Error: expect(@miq_server2.active_roles.size).to eq 3

       expected: 3
            got: 4

       (compared using ==)
     # ./spec/models/miq_server/server_monitor_spec.rb:334:in `block (5 levels) in <top (required)>'

  2) Server Monitor After Setup, with 2 Servers in 2 Zones where I am the Master should evenly distribute singular roles on multiple servers with identical assigned roles
     Failure/Error: expect(@miq_server1.active_roles.count).to eq(@miq_server2.active_roles.count)

       expected: 4
            got: 2

       (compared using ==)
     # ./spec/models/miq_server/server_monitor_spec.rb:294:in `block (5 levels) in <top (required)>'

  3) Server Monitor After Setup, with 2 Servers in 2 Zones where I am the Master should evenly distribute singular roles on multiple servers with partially shared assigned roles
     Failure/Error: expect(@miq_server1.active_role_names).to match_array(%w(event notifier storage_inventory ems_metrics_coordinator))

       expected collection contained:  ["ems_metrics_coordinator", "event", "notifier", "storage_inventory"]
       actual collection contained:    ["ems_metrics_coordinator", "event", "notifier", "scheduler", "storage_inventory"]
       the extra elements were:        ["scheduler"]
     # ./spec/models/miq_server/server_monitor_spec.rb:304:in `block (5 levels) in <top (required)>'
```

EDIT:

Caveats: 
* each role is not created equal so this only prevents singular roles from sticking to one server, not to truly even the load on the appliances
* we don't shift roles after activation, we should probably have a way of doing this on servers that are too busy and there are more idle servers to help

jrafanie commented:
@Fryguy @gtanzillo @blomquisg please review

jrafanie commented:
EDIT, see updated description.

jrafanie commented on app/models/miq_server/role_management.rb:
```diff
@@ -239,6 +239,16 @@ def synchronize_active_roles(servers, roles_to_sync)
+              # To do this in ruby, we do it in reverse order so the most important
+              # sort, priority(affinity), happens last.
+              inactive = inactive
+                .sort { |a, b| a.first.reload.active_roles.length <=> b.first.reload.active_roles.length }
```
We could track all of the roles we activate in this method but it's a lot of bookkeeping.  We can do that if this reloading of the server is expensive.

Fryguy commented on app/models/miq_server/role_management.rb:
```diff
@@ -239,6 +239,16 @@ def synchronize_active_roles(servers, roles_to_sync)
+              # sort, priority(affinity), happens last.
+              inactive = inactive
+                .sort { |a, b| a.first.reload.active_roles.length <=> b.first.reload.active_roles.length }
+                .sort { |a, b| a.last <=> b.last }
```
This is crazy complicated, plus that reload in the sort block will be crazy expensive because it will re-reload the same server objects over and over (perfect use case for sort_by)... how about:

```ruby
inactive.sort_by do |server, priority|
  [priority, -server.reload.active_roles.size]
end
```

You can just toss `-` negatives into whichever value to get right ascend/descend sorting pair...I'm not sure how priority is represented.

jrafanie replied:
Cool, yeah, good idea.  👍 

I can try it out, I just threw up the idea to see what people thought of the concept.

jrafanie replied:
Close @Fryguy, see updated change, I don't need to use `-` since sort_by returns the numbers in ascending order, which is what we want for both priority (lower is higher) and roles (servers with less roles need to come first)

jrafanie commented on app/models/miq_server/role_management.rb:
```diff
@@ -239,6 +239,13 @@ def synchronize_active_roles(servers, roles_to_sync)
+              inactive = inactive.sort_by do |server, priority|
+                [priority, server.reload.active_roles.size]
+              end
+
```
Note, we probably have to do something similar on line 251/258 below when it migrates roles from one server to another.  It needs to choose the destination server based on priority (affinity) AND the number of roles active on that server.

👇

jrafanie replied:
I'm all 👂  to suggestions on extracting this logic

chrisarcand replied:
Thought about this a while for fun but I think any extraction does more harm than good just because of the fact that you're dealing with local vars in this context and not actual attrs on the sortables (I think).

jrafanie replied:
@chrisarcand Yes, it felt weird duplicating the logic but I agree, it seems like logic very local to this method so I resisted the temptation to DRY it up.

gtanzillo commented:
@jrafanie I definitely like this approach 👍  What you have so far looks good. I would think similar logic would be necessary for migrating the roles.

jrafanie commented:
Ok, @Fryguy @gtanzillo, updated to include all suggestions and also added the sort for the migration of active roles from lower priority assigned servers to higher priority ones.  I spent much more time on the tests than the code 😉

jrafanie commented:
Closing for now as I try tackling this from a different angle.
